accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+/**
+ * AccessibleBlockly
+ *
+ * Copyright 2016 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Angular2 Service that handles all tree keyboard navigation.
+ * @author madeeha@google.com (Madeeha Ghori)
+ */
+blocklyApp.TreeService = ng.core
+  .Class({
+    constructor: function() {
+      console.log('making a new tree service');
+      this.activeDesc_ = {};
+      this.previousKey_;
+      this.trees = document.getElementsByClassName('blocklyTree');
+    },
+    createId: function(obj) {
+      if (obj && obj.id) {
+        return obj.id;
+      }
+      return 'blockly-' + Blockly.genUid();
+    },
+    setActiveAttribute: function(tree) {
+      if (!document.getElementById(tree.getAttribute(
+          'aria-activedescendant'))) {
+        blocklyApp.debug && console.log('setting tree active descendant');
+        tree.setAttribute('aria-activedescendant', tree.id + '-node0');
+      }
+    },
+    setActiveDesc: function(node, id) {
+      blocklyApp.debug && console.log('setting active descendant for tree ' + id);
+      this.activeDesc_[id] = node;
+    },
+    getActiveDesc: function(id) {
+      return this.activeDesc_[id];
+    },
+    updateSelectedNode: function(node, tree, keepFocus) {
+      blocklyApp.debug && console.log('updating node: ' + node.id);
+      var treeId = tree.id;
+      var activeDesc = this.getActiveDesc(treeId);
+      if (activeDesc) {
+        activeDesc.classList.remove('blocklyActiveDescendant');
+      } else {
+        blocklyApp.debug && console.log('updateSelectedNode: there is no active descendant');
+      }
+      node.classList.add('blocklyActiveDescendant');
+      tree.setAttribute('aria-activedescendant', node.id);
+      this.setActiveDesc(node, treeId);
+      node.setAttribute('aria-selected', 'true');
+
+      // make sure keyboard focus is on tree as a whole
+      // in case before the user was editing a block and keyboard
+      // focus got shifted.
+      if (keepFocus) {
+        tree.focus();
+      }
+    },
+    workspaceButtonKeyHandler: function(e, treeId) {
```

Comment:
The main idea I was going for is that function names should start with a verb, since they 'do' something. E.g. paste() (since that's what the function does when you call it) rather than pasteHandler() (which describes the function).

Another convention I've seen is onWorkspaceKeypress(). That works too.